--- a/ext/exif/exif.c	2010-01-03 09:23:27.000000000 +0000
+++ b/ext/exif/exif.c	2015-10-21 16:49:26.227209421 +0000
@@ -2441,11 +2441,11 @@ static void* exif_ifd_make_value(image_i
 					data_ptr += 8;
 					break;
 				case TAG_FMT_SINGLE:
-					memmove(data_ptr, &info_data->value.f, byte_count);
+					memmove(data_ptr, &info_value->f, 4);
 					data_ptr += 4;
 					break;
 				case TAG_FMT_DOUBLE:
-					memmove(data_ptr, &info_data->value.d, byte_count);
+					memmove(data_ptr, &info_value->d, 8);
 					data_ptr += 8;
 					break;
 			}

