--- a/ext/mysqli/mysqli_api.c	2013-12-10 19:04:57.000000000 +0000
+++ b/ext/mysqli/mysqli_api.c	2015-10-21 15:08:29.915433295 +0000
@@ -1830,7 +1830,10 @@ PHP_FUNCTION(mysqli_prepare)
 		efree(stmt);
 		RETURN_FALSE;
 	}
-
+#ifndef MYSQLI_USE_MYSQLND
+	stmt->link_handle = Z_OBJ_HANDLE(*mysql_link);
+	zend_objects_store_add_ref_by_handle(stmt->link_handle TSRMLS_CC);
+#endif
 	mysqli_resource = (MYSQLI_RESOURCE *)ecalloc (1, sizeof(MYSQLI_RESOURCE));
 	mysqli_resource->ptr = (void *)stmt;
 
@@ -2358,6 +2361,10 @@ PHP_FUNCTION(mysqli_stmt_init)
 		efree(stmt);
 		RETURN_FALSE;
 	}
+#ifndef MYSQLI_USE_MYSQLND
+       stmt->link_handle = Z_OBJ_HANDLE(*mysql_link);
+       zend_objects_store_add_ref_by_handle(stmt->link_handle TSRMLS_CC);
+#endif
 
 	mysqli_resource = (MYSQLI_RESOURCE *)ecalloc (1, sizeof(MYSQLI_RESOURCE));
 	mysqli_resource->status = MYSQLI_STATUS_INITIALIZED;
--- a/ext/mysqli/mysqli.c	2013-12-10 19:04:57.000000000 +0000
+++ b/ext/mysqli/mysqli.c	2015-10-21 15:10:15.367429397 +0000
@@ -174,8 +174,10 @@ void php_clear_stmt_bind(MY_STMT *stmt T
 	php_free_stmt_bind_buffer(stmt->param, FETCH_SIMPLE);
 	/* Clean output bind */
 	php_free_stmt_bind_buffer(stmt->result, FETCH_RESULT);
+	if (stmt->link_handle) {
+		zend_objects_store_del_ref_by_handle(stmt->link_handle TSRMLS_CC);
+	}
 #endif
-
 	if (stmt->query) {
 		efree(stmt->query);
 	}
@@ -1020,6 +1022,10 @@ PHP_FUNCTION(mysqli_stmt_construct)
 		efree(stmt);
 		RETURN_FALSE;
 	}
+#ifndef MYSQLI_USE_MYSQLND
+	stmt->link_handle = Z_OBJ_HANDLE(*mysql_link);
+	zend_objects_store_add_ref_by_handle(stmt->link_handle TSRMLS_CC);
+#endif
 
 	mysqli_resource = (MYSQLI_RESOURCE *)ecalloc (1, sizeof(MYSQLI_RESOURCE));
 	mysqli_resource->ptr = (void *)stmt;
--- a/ext/mysqli/php_mysqli_structs.h	2013-12-10 19:04:57.000000000 +0000
+++ b/ext/mysqli/php_mysqli_structs.h	2015-10-21 15:12:41.619423991 +0000
@@ -114,6 +114,9 @@ typedef struct {
 	BIND_BUFFER	param;
 	BIND_BUFFER	result;
 	char		*query;
+#ifndef MYSQLI_USE_MYSQLND
+	zend_object_handle	link_handle;
+#endif
 } MY_STMT;
 
 typedef struct {

